In android builds and linux tidy:

```
merror[E0063]: missing field `xr_discovery` in initializer of `simpleservo::InitOptions`
   --> ports/libsimpleservo/capi/src/lib.rs:136:16
    |
136 |     let opts = InitOptions {
    |                m^^^^^^^^^^^ mmissing `xr_discovery`

merror: aborting due to previous error

For more information about this error, try `rustc --explain E0063`.
error: Could not compile `simpleservo_capi`.
```

--- a/ports/libmlservo/Cargo.toml
+++ b/ports/libmlservo/Cargo.toml
@@ -23,6 +23,8 @@ layout-2020 = ["simpleservo/layout-2020"]
 libservo = { path = "../../components/servo", features = ["no_static_freetype"] }
 simpleservo = { path = "../libsimpleservo/api", features = ["no_static_freetype"] }
 rust-webvr = { version = "0.13", features = ["magicleap"] }
+webxr-api = { git = "https://github.com/servo/webxr", features = ["ipc"] }
+webxr = { git = "https://github.com/servo/webxr", features = ["ipc", "magicleap"] }
 libc = "0.2"
 log = "0.4"
 servo-egl = "0.2"

--- a/ports/libmlservo/src/lib.rs
+++ b/ports/libmlservo/src/lib.rs
@@ -29,6 +29,7 @@ use std::rc::Rc;
 use std::thread;
 use std::time::Duration;
 use std::time::Instant;
+use webxr::magicleap::MagicLeapDiscovery;
 
 #[repr(u32)]
 pub enum MLLogLevel {
@@ -106,7 +107,7 @@ pub unsafe extern "C" fn init_servo(
     url_update: MLURLUpdate,
     keyboard: MLKeyboard,
     url: *const c_char,
-    args: *const c_char,
+    default_args: *const c_char,
     width: u32,
     height: u32,
     hidpi: f32,
@@ -117,7 +118,6 @@ pub unsafe extern "C" fn init_servo(
 
     let gl = gl_glue::egl::init().expect("EGL initialization failure");
 
-    let url = CStr::from_ptr(url).to_str().unwrap_or("about:blank");
     let coordinates = Coordinates::new(
         0,
         0,
@@ -126,34 +126,52 @@ pub unsafe extern "C" fn init_servo(
         width as i32,
         height as i32,
     );
-    let args = if args.is_null() {
-        vec![]
-    } else {
-        CStr::from_ptr(args)
+
+    let mut url = CStr::from_ptr(url).to_str().unwrap_or("about:blank");
+
+    // If the URL has a space in it, then treat everything before the space as arguments
+    let args = if let Some(i) = url.rfind(' ') {
+        let (front, back) = url.split_at(i);
+        url = back;
+        front.split(' ').map(|s| s.to_owned()).collect()
+    } else if !default_args.is_null() {
+        CStr::from_ptr(default_args)
             .to_str()
             .unwrap_or("")
             .split(' ')
             .map(|s| s.to_owned())
             .collect()
+    } else {
+        Vec::new()
     };
+
     info!("got args: {:?}", args);
 
-    let vr_init = if landscape {
-        VRInitOptions::None
-    } else {
+    let vr_init = if !landscape {
         let name = String::from("Magic Leap VR Display");
         let (service, heartbeat) = MagicLeapVRService::new(name, ctxt, gl.gl_wrapper.clone())
             .expect("Failed to create VR service");
         let service = Box::new(service);
         let heartbeat = Box::new(heartbeat);
         VRInitOptions::VRService(service, heartbeat)
+    } else {
+        VRInitOptions::None
     };
+
+    let xr_discovery: Option<Box<dyn webxr_api::Discovery>> = if !landscape {
+        let discovery = MagicLeapDiscovery::new(ctxt, gl.gl_wrapper.clone());
+        Some(Box::new(discovery))
+    } else {
+        None
+    };
+
     let opts = InitOptions {
         args,
         url: Some(url.to_string()),
         density: hidpi,
         enable_subpixel_text_antialiasing: false,
         vr_init,
+        xr_discovery,
         coordinates,
         gl_context_pointer: Some(gl.gl_context),
         native_display_pointer: Some(gl.display),

--- a/ports/libsimpleservo/api/Cargo.toml
+++ b/ports/libsimpleservo/api/Cargo.toml
@@ -10,6 +10,7 @@ publish = false
 libservo = { path = "../../../components/servo" }
 log = "0.4"
 servo-media = { git = "https://github.com/servo/media" }
+webxr-api = { git = "https://github.com/servo/webxr" }
 
 [target.'cfg(not(target_os = "macos"))'.dependencies]
 libc = "0.2"

--- a/ports/libsimpleservo/api/src/lib.rs
+++ b/ports/libsimpleservo/api/src/lib.rs
@@ -48,6 +48,7 @@ pub struct InitOptions {
     pub coordinates: Coordinates,
     pub density: f32,
     pub vr_init: VRInitOptions,
+    pub xr_discovery: Option<Box<dyn webxr_api::Discovery>>,
     pub enable_subpixel_text_antialiasing: bool,
     pub gl_context_pointer: Option<*const c_void>,
     pub native_display_pointer: Option<*const c_void>,
@@ -198,6 +199,7 @@ pub fn init(
 
     let embedder_callbacks = Box::new(ServoEmbedderCallbacks {
         vr_init: init_opts.vr_init,
+        xr_discovery: init_opts.xr_discovery,
         waker,
     });
 
@@ -576,6 +578,7 @@ impl ServoGlue {
 
 struct ServoEmbedderCallbacks {
     waker: Box<dyn EventLoopWaker>,
+    xr_discovery: Option<Box<dyn webxr_api::Discovery>>,
     vr_init: VRInitOptions,
 }
 
@@ -604,6 +607,13 @@ impl EmbedderMethods for ServoEmbedderCallbacks {
                 services.register(service);
                 heartbeats.push(heartbeat);
             },
+        }
+    }
+
+    fn register_webxr(&mut self, registry: &mut webxr_api::MainThreadRegistry) {
+        debug!("EmbedderMethods::register_xr");
+        if let Some(discovery) = self.xr_discovery.take() {
+            registry.register(discovery);
         }
     }
 

--- a/ports/libsimpleservo/capi/src/lib.rs
+++ b/ports/libsimpleservo/capi/src/lib.rs
@@ -143,6 +143,7 @@ unsafe fn init(
         } else {
             VRInitOptions::VRExternal(opts.vr_pointer)
         },
+        xr_discovery: None,
         enable_subpixel_text_antialiasing: opts.enable_subpixel_text_antialiasing,
         gl_context_pointer: gl_context,
         native_display_pointer: display,

--- a/ports/libsimpleservo/jniapi/src/lib.rs
+++ b/ports/libsimpleservo/jniapi/src/lib.rs
@@ -745,6 +745,7 @@ fn get_options(env: &JNIEnv, opts: JObject) -> Result<(InitOptions, bool, Option
         } else {
             VRInitOptions::VRExternal(vr_pointer)
         },
+        xr_discovery: None,
         gl_context_pointer: None,
         native_display_pointer: None,
     };

